ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm

```diff
@@ -0,0 +1,146 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import "CKCollectionViewTransactionalDataSource.h"
+
+#import "CKCollectionViewDataSourceCell.h"
+#import "CKTransactionalComponentDataSourceConfiguration.h"
+#import "CKTransactionalComponentDataSourceListener.h"
+#import "CKTransactionalComponentDataSourceItem.h"
+#import "CKTransactionalComponentDataSourceState.h"
+#import "CKTransactionalComponentDataSourceAppliedChanges.h"
+#import "CKComponentRootView.h"
+#import "CKComponentLayout.h"
+#import "CKComponentDataSourceAttachController.h"
+
+@interface CKCollectionViewTransactionalDataSource () <
+UICollectionViewDataSource,
+CKTransactionalComponentDataSourceListener
+>
+{
+  CKTransactionalComponentDataSource *_componentDataSource;
+  CKTransactionalComponentDataSourceState *_currentState;
+  CKComponentDataSourceAttachController *_attachController;
+}
+@end
+
+@implementation CKCollectionViewTransactionalDataSource
+
+- (instancetype)initWithCollectionView:(UICollectionView *)collectionView
+                         configuration:(CKTransactionalComponentDataSourceConfiguration *)configuration
+{
+  self = [super init];
+  if (self) {
+    _componentDataSource = [[CKTransactionalComponentDataSource alloc] initWithConfiguration:configuration];
+    [_componentDataSource addListener:self];
+      
+    _collectionView = collectionView;
+    _collectionView.dataSource = self;
+    [_collectionView registerClass:[CKCollectionViewDataSourceCell class] forCellWithReuseIdentifier:kReuseIdentifier];
+    
+    _attachController = [[CKComponentDataSourceAttachController alloc] init];
+  }
+  return self;
+}
+
+#pragma mark - Changeset application
+
+- (void)applyChangeset:(CKTransactionalComponentDataSourceChangeset *)changeset
+                  mode:(CKTransactionalComponentDataSourceMode)mode
+              userInfo:(NSDictionary *)userInfo
+{
+  [_componentDataSource applyChangeset:changeset
+                                  mode:mode
+                              userInfo:userInfo];
+}
+
+static void applyChangesToCollectionView(CKTransactionalComponentDataSourceAppliedChanges *changes, UICollectionView *collectionView)
+{
+  [collectionView reloadItemsAtIndexPaths:[changes.updatedIndexPaths allObjects]];
```

Comment:
(Certainly not a regression since the existing open source DS does it this way too)